xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,290 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
   EXPECT_FALSE(IsAsync(all_reduce));
 }
 
+// E2E tests comparing the results of sharded and unsharded execution.
+class CollectiveOpsTestE2EShardedUnsharded : public CollectiveOpsTestE2E {
+ public:
+  void CollectiveOpsCompareShardedUnsharded(const std::string& hlo_text,
+                                            const int64_t num_partitions = 2) {
+    const int64_t num_replicas = 1;
+    if (test_runner().device_count() < num_replicas * num_partitions) {
+      GTEST_SKIP() << "Test requires at least " << num_replicas * num_partitions
+                   << " devices (" << test_runner().device_count()
+                   << " available)";
+    }
+
+    TF_ASSERT_OK_AND_ASSIGN(std::vector<Literal> ref_results,
+                            ExecuteReplicated(hlo_text));
+    ASSERT_EQ(ref_results.size(), 1);
+
+    TF_ASSERT_OK_AND_ASSIGN(
+        std::vector<Literal> results,
+        ExecuteSharded(hlo_text, num_partitions, num_replicas));
+    ASSERT_EQ(results.size(), num_partitions);
+
+    // Slice the unsharded reference results and compare to the sharded case.
+    HloModuleConfig config =
+        GetModuleConfigForTest(/*replica_count=*/num_replicas);
+    auto opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSERT_OK_AND_ASSIGN(auto module,
+                            ParseAndReturnVerifiedModule(hlo_text, config));
```

Comment:
Why parse the module here again? Could you just pass the needed info into the test case?

Reply:
Passing the module into the functions isn't ideal since ParseAndReturnVerifiedModule returns a unique_ptr. Let me know if you want to change this anyway.